pdbApp/pdbgroup.cpp

```diff
@@ -349,12 +350,24 @@ void PDBGroupPut::put(pvd::PVStructure::shared_pointer const & value,
     // assume value may be a different struct each time... lot of wasted prep work
     const size_t npvs = channel->pv->members.size();
     std::vector<std::tr1::shared_ptr<PVIF> > putpvif(npvs);
+    std::vector<AsWritePvt> asWritePvt;
 
     for(size_t i=0; i<npvs; i++)
     {
         PDBGroupPV::Info& info = channel->pv->members[i];
-        if(!info.allowProc) continue;
 
+        asWritePvt.push_back(AsWritePvt(
```

Comment:
Ah, I hadn't thought through all of the implications when providing a prospective definition for `AsWritePvt`.  Specifically, that definition can't be safely copied.  Using `push_back()` is making a copy, and so strictly speaking (with c++98) the destructor should run twice, and result in a double free error.

The fact that it does work in many cases is due to compilers (GCC and recent MSVC) eliding the copy.  However, this is not required.

Since I'm not (yet) requiring c++11 support, it isn't possible to use a move constructor.  An older way to handle this is with `swap()` as can be seen with `class DBCH` in pvif.h (which wraps a `dbChannel*`).

```c++
class AsWritePvt {
    void * pvt;
public:
    AsWritePvt() :pvt(0) {}
    explicit AsWritePvt(void * pvt): pvt(pvt) {}
    ~AsWritePvt() {
        if(pvt) asTrapWriteAfterWrite(pvt);
    }
    void swap(AsWritePvt& o) {
        std::swap(pvt, o.pvt);
    }
private:
    AsWritePvt(const AsWritePvt&);
    AsWritePvt& operator=(const AsWritePvt&);
};
```

Usage is then a bit awkward.

```c++
AsWritePvt temp(...);
asWritePvt.push_back(AsWritePvt());
asWritePvt.back().swap(temp);
```

Reply:
I had to make the copy constructor public otherwise `asWritePvt.push_back(AsWritePvt());` fails as it cannot make a copy (of an empty object). I am setting the pointer to `NULL` with copy constructor (and I made it explicit). 

Does this look good to you?
```
class AsWritePvt {
    void * pvt;
public:
    AsWritePvt() :pvt(0) {}
    explicit AsWritePvt(const AsWritePvt&) : pvt(NULL) {}
    explicit AsWritePvt(void * pvt): pvt(pvt) {}
    ~AsWritePvt() {
        if(pvt) asTrapWriteAfterWrite(pvt);
    }
    void swap(AsWritePvt& o) {
        std::swap(pvt, o.pvt);
    }
private:
    AsWritePvt& operator=(const AsWritePvt&);
};
```

Thanks

Reply:
> I had to make the copy constructor public otherwise asWritePvt.push_back(AsWritePvt()); fails

Right, which is why `std::vector` can't be used to store non-copyable types until `std::move()` is available.  Part of the solution is to use `shared_vector`, which has an additional constructor which will default construct each element.  This avoids the need for copying.

>  ... I am setting the pointer to NULL with copy constructor (and I made it explicit).

Various `std::` container classes, including `std::vector` make temporary copies during some operations.  Having a copy constructor which doesn't copy will cause these to break is non-obvious ways.  cf. discussions of the chaos caused by attempts to use `std::vector<std::auto_ptr>`.